card/mod/core/set/all/fetch.rb

```diff
@@ -246,21 +246,33 @@ def normalize_fetch_args args
 
   def normalize_mark mark, opts
     case mark
-    when Symbol  then Card::Codename[mark]
-    when Integer then mark.to_i
-    when Card    then mark.cardname
-    when String, SmartName
+    when Symbol            then require_id_for_codename mark
+    when Integer           then mark.to_i
+    when Card              then mark.cardname
+    when String, SmartName then normalize_stringy_mark mark, opts
       # there are some situations where this breaks if we use Card::Name
       # rather than SmartName, which would seem more correct.
       # very hard to reproduce, not captured in a spec :(
-      case mark.to_s
-      when /^\~(\d+)$/ then $1.to_i                   # id
-      when /^\:(\w+)$/ then Card::Codename[$1.to_sym] # codename
-      else fullname_from_mark mark, opts[:new]        # name
-      end
     end
   end
 
+  def normalize_stringy_mark mark, opts
+    case mark.to_s
+    when /^\~(\d+)$/  # id, eg "~75"
+      Regexp.last_match[1].to_i
+    when /^\:(\w+)$/  # codename, eg ":options"
+      require_id_for_codename(Regexp.last_match[1].to_sym)
+    else
+      fullname_from_mark mark, opts[:new]
+    end
+  end
+
+  def require_id_for_codename mark
+    id = Card::Codename[mark]
+    raise Card::NotFound, "missing card with codename: #{mark}" unless id
```

Comment:
is `Card::NotFound` right? in CardController#load_card we raise `Card::Error::NotFound`